src/Psalm/Internal/Analyzer/Statements/Expression/Call/FunctionCallAnalyzer.php

```diff
@@ -1034,4 +1035,42 @@ private static function checkFunctionCallPurity(
             }
         }
     }
+
+    private static function callUsesByReferenceArguments(
+        FunctionCallInfo $function_call_info,
+        PhpParser\Node\Expr\FuncCall $stmt
+    ): bool {
+        // If the function doesn't have any by-reference parameters
+        // we shouldn't look any further.
+        if (!$function_call_info->hasByReferenceParameters() || null === $function_call_info->function_params) {
+            return false;
+        }
+
+        $parameters = $function_call_info->function_params;
+
+        // If no arguments were passed
+        if (0 === \count($stmt->args)) {
+            return false;
+        }
+
+        foreach ($stmt->args as $index => $argument) {
+            if (null !== $argument->name) {
+                $argument_name = $argument->name->toString();
+                foreach ($parameters as $param) {
+                    if ($param->name === $argument_name) {
+                        $parameter = $param;
+                        break;
+                    }
+                }
+            } else {
+                $parameter = $parameters[$index] ?? null;
+            }
+
+            if ($parameter && $parameter->by_ref) {
+                return true;
+            }
+        }
+
+        return false;
+    }
```

Comment:
something doesn't feel right about this function here, i added as many tests as i could have thought of, all seem to be passing as expected, but need further review.